### DISCUSSION TITLE
Why does A come in as a key press but @ come in as a Text Event

### DISCUSSION THREAD

**Discussion by pm100** [emilk/egui#4703] [General]
Lower case 'a' is reported in input.keys_down
shift + 'a' is reported in input.keys_down + modifier = shift 
Number '2' is reported in keys_down '2'
Number 2 + shift is reported as a Text event = "@"

Most (not all) of the non alpha keys when shifted are reported as a Text event

This is 
- very confusing and hard to debug
- not documented
- fiddly to code with

**Comment by emilk** [comment]
On what platform? For me, using `eframe/winit` (native) I get proper key events and text events for all these cases. You can use the new `Input Event History` in the egui demo app:

![image](https://github.com/emilk/egui/assets/1148717/ce79a72a-07d0-435a-b80c-43cdf0271009)
(read bottom->up)

**Comment by pm100** [comment]
windows, using egui_template app.

I have this code at top of my update

```
        if !ctx.wants_keyboard_input() {
            ctx.input(|inp| {
                // log::trace!("{:?}", inp);
                if inp.keys_down.len() > 0 {
                    println!(
                        "{:?} {:?} {} ",
                        inp.keys_down,
                        inp.modifiers,
                        inp.keys_down.len()
                    );
                    lookup_key(inp);
                    println!("key: {}", unsafe { CURRENT_KEY as char });
                } else {
                    unsafe {
                        CURRENT_KEY = 0;
                    }
                }
                if let Some(ev) = inp.events.iter().next() {
                    if let egui::Event::Text(text) = ev {
                        println!("{:?}", text);
                        unsafe {
                            CURRENT_KEY = text.chars().next().unwrap() as u8;
                        }
                    }
                }
            });
        }

```

I get

```
{Num2} Modifiers { alt: false, ctrl: false, shift: false, mac_cmd: false, command: false } 1
key: 2
{Num2} Modifiers { alt: false, ctrl: false, shift: false, mac_cmd: false, command: false } 1
key: 2
{Num2} Modifiers { alt: false, ctrl: false, shift: false, mac_cmd: false, command: false } 1
key: 2
"@"
```

when I press 2 and then shift + 2

The first three come from the 2 key being down, the @ comes when I do shift then 2

However your demo app shows the correct thing for me!

code is here https://github.com/pm100/hackem

Just cargo run it and type (ignore ui and menus etc)

**Reply by emilk** [reply]
Have you tried latest egui master?

**Reply by pm100** [reply]
NOTE - I started using master but had to revert, egui_extras wont compile with egui feature persistence selected. And I need images (and not having persistence sucks). I know how to live with this text input error for now

**Comment by pm100** [comment]
OK using egui / eframe master . Now all keys come as keys_down. Great!
However no longer getting Text events, which is a surprise since thats a breaking change
